xcept/xcept.1
.TH XCEPT 1 "24 September 1993"
.SH NAME
xcept \- a Bildschirmtext-decoder for the X\-Window system
.SH SYNOPSIS
.B xcept
[\-h host] [\-p port] [\-l file] [\-t file] [\-r file] [\-c] [\-j count]
[\-x] [\-X] [\-b] [\-2] [\-m] [\-S pathlist] [\-s file [parameter]] 
.SH DESCRIPTION
.I xcept
is a Bildschirmtext terminal for the CEPT-service of the Deutsche Bundespost
Telekom.
.I xcept
creates a window on the X-display and then connects to the
.I ceptd.
After the
.I ceptd
established the connection to the CEPT-exchange,
.I xcept
is ready to exchange data.
.PP
The following options are available:
.PP
.TP
.B \-h host
Connect to ceptd at host ``host''.
.TP
.B \-p port
Use socket port number ``port'' for connection to ceptd.
.TP
.B \-l file
Log CEPT-session to file ``file''.
.TP
.B \-t file
Log debug text to file ``file''.
.TP
.B \-r file
Replay CEPT-session from file ``file''. Stops at every page \-
press any key to continue.
.TP
.B \-j count
Jump over the first ``count'' pages (playback mode only).
.TP
.B \-c
Connect to ceptd automatically on startup.
.TP
.B \-x
Don't open X-window on startup (you have to run a script).
.TP
.B \-X
Turn on synchronous mode for Xlib calls. Useful for debugging.
.TP
.B \-b
Don't use Xserver's backingstore capability.
.TP
.B \-2
Double size. Display each pixel as 2x2 pixels.
.TP
.B \-m
Force monochrome display.
.TP
.B \-S pathes
Specify a colon separated list of pathes where to look for .xcept script
files.
.TP
.B \-s file [parameter]
Run script ``file'' on startup with optional parameters. This has to be the
last option. All following arguments are treated as script parameters
(script variables
.B argv_1, argv_2, ...
).
.PP
In addition to all alphanumeric keyboard symbols and cursor keys the
following special keyboard commands are valid:
.PP
.TP
.B 'ESC'
Exit from xcept immediately.
.TP
.B '^L'
Redraw the CEPT window.
.TP
.B '^Q???'
3-digit octal quote. Send the specified 8-bit character code.
.TP
.B '*'
Send the CEPT initiator character.
.TP
.B '#', 'RETURN'
Send the CEPT terminator character.
.TP
.B Modifier 1 Key (ALT) + [aAoOuUs]
Send german 'a umlaut', 'A umlaut', 'o umlaut', 'O umlaut', 'u umlaut', 'U umlaut', and 'sharp s' character.
.TP
.B '^W' 
Append current page as text to a file. The filename requester will be
popped up only once.
.PP
.SH XCEPT SCRIPT LANGUAGE
The language works on variables or string constants. Variable names can be of
any length and consist of the characters [A-Z0-9a-z_]. String constants are
enclosed in double quotes and may contain any ASCII characters.
Script parameters are named
.B argv_0
through
.B argv_n.
.B argv_0
is the name
of the script file. The number of arguments can be found in the
variable named
.B argc.
Any characters following a sharp sign (#) until the end of the line are
treated as a comment.
If the special script variable "DEBUG" is set to "on", each executed
instruction is printed to stderr.
.PP
Special script string characters are:
.TP
.B \eb
Write a <backspace> character.
.TP
.B \ef
Write a <form-feed> character.
.TP
.B \en
Write a <new-line> character.
.TP
.B \er
Write a <carriage return> character.
.TP
.B \et
Write a <tab> character.
.TP
.B \ev
Write a <vertical tab> character.
.TP
.B \e\e
Write a <backslash> character.
.TP
.B \enum
Write an 8-bit character whose ASCII value is the 1-, 2-,
or 3-digit octal number num.
.TP
.B \ea
Write a german <a umlaut> character.
.TP
.B \eA
Write a german <A umlaut> character.
.TP
.B \eo
Write a german <o umlaut> character.
.TP
.B \eO
Write a german <O umlaut> character.
.TP
.B \eu
Write a german <u umlaut> character.
.TP
.B \eU
Write a german <U umlaut> character.
.TP
.B \es
Write a german <sharp s> character.
.TP
.B \e#
Write the CEPT <terminator> character.
.TP
.B \e*
Write the CEPT <initiator> character.
.PP
.SH SCRIPT COMMANDS
.TP
set <var A>=<var B|str B>
Assign value of B to variable A.
.TP
write <var A|str A> to <var B|str B>
Write contents of A to file B.
.TP
read <var A> from <var B|str B>
Read contents of file B into variable A.
.TP
append <var A|str A> to <var B|str B>
Append contents of A to file B.
.TP
strcat <var A> , <var B|str B>
Concatenate B to A.
.TP
send <var A|str A>
Send A to the CEPT-exchange.
.TP
get <var A>=<var B|str B>,<var C|str C>,<var D|str D>,
<var E|str E>
.br
Get the contents of the screen rectangle specified by B, C, D, E into
variable A. B, C specify the coordinates of the upper left corner
and D, E the coordinates of the lower right corner of the rectangle.
The coordinates are 1 based (x-range: 1-40, y-range:1-24).
.TP
waitdct
Wait for completion of the next page or user input.
.TP
if <var A|str A> == <var B|str B> goto <label C>
Conditional goto. If A is equal to B then continue execution at instruction
labeled C. Label names are defined with a following colon (:), but referenced
without one. (forever:  goto forever)
.TP
info <var A|str A>
Write contents of A to <stdout>.
.TP
debug <var A|str A>
Write contents of A to <stderr>.
.TP
goto <label A>
Unconditional goto to label A.
.TP
inc <var A>
Increment the numerical value of variable A.
.TP
dec <var A>
Decrement the numerical value of variable A.
.TP
setpart <var A>=<var B|str B>,<var C|str C>,<var D|str D> 
Copy the contents of B beginning from position C up to position D into
variable A.
.TP
system <var A|str A>
Execute shell command A.
.TP
sleep <var A|str A>
Sleep A seconds.
.TP
onx
Open the xcept X-window at the display specified in the script variable
"DISPLAY" or at the default display (DISPLAY environment variable).
.TP
offx
Close the xcept X-window.
.TP
input <var A> :  <var B|str B> , <var C|str C>
Popup a requester and get user input into variable A.
Prompt string is B and default answer
string is C. If C is equal to "*NO ECHO*", the user input isn't
echoed to the window (useful for password input).
If C is equal to "*STDIN*", input is read from stdin and not from a
requester (useful for piped input).
.TP
connect
Connect to the ceptd.
.TP
disconnect
Close the connection to the ceptd.
.TP
playback <var A|str A>
Start session playback from file A.
.TP
ppmsave <var A|str A>
Save current CEPT page as PPM image to file A.
.TP
getstate <var A>
Get the current terminal state into variable A. Possible states are
"CONNECTED", "ONLINE", "DISCONNECTED" or "PLAYBACK".
.TP
quit
Exit from xcept immediately.

.SH EXAMPLE
This is a sample script (login.scp) which logs into CEPT as guest and selects
page 50000.
.PP
Start with: xcept -s login.scp 50000
.PP
.RS +.8i
connect
.br
waitdct                   # wait for login page
.br
send       "\e#"           # accept guest id
.br
waitdct
.br
set        page = "\e*"
.br
strcat     page, argv_1
.br
strcat     page, "\e#"
.br
send       page           # '*50000#'
.br
waitdct
.RE
.SH ENVIRONMENT
.PP
.TP
XCEPTSCRIPTS 
overwrites the default search path for script files.
(e.g. ~/.xcept:/local/lib/xcept)
.TP
XCEPTHOST 
overwrites the default hostname of server host
(e.g. myhost.informatik.uni-erlangen.de)
.TP 
For defaults type xcept -h.
.RE
.SH "SEE ALSO"
ceptd(8)
.br
"Functional Specification for BTX-Terminals", DBP Telekom

.SH AUTHORS
.TP
Arno Augustin, University of Erlangen-Nuremberg,
(aoaugust@cip.informatik.uni-erlangen.de)
.TP
Frank Hoering, University of Erlangen-Nuremberg,
(fkhoerin@cip.informatik.uni-erlangen.de)

.SH BUGS
Not all features are documented.
For further information read the source code.


lib/xcept.pl
# perl adapter to xcept
# by Thomas Eirich (eirich@informatik.uni-erlangen.de)
#

package xcept;

require 'sys/socket.ph';
require 'fcntl.ph';

$BTX           = "/usr/local/btx";		# path to btx installation
$prog_pid      = 0;			# xcept's pid if started successfully
$error_msg     = "";			# error strings if commands have failed
@reply	       = ();			# reply to a cmd
$prompt        = "XCEPT??????\n";	# xcept prompt string
$prefix        = "XCEPT!!!!!!";		# xcept prefix for non error messages
$version       = 1.0;			# expected version of xcept's client mode
$errlevel      = 0;			# flag for error in error handling

################################################################################
##  the following variables may be changed prior to calling
##  initiate() to change some default behavior.

$prog        = "$BTX/bin/xcept";	# the xcept server executable
$prog_opts   = "-x";			# more program options (like -h etc.)
$client_mode = "$BTX/lib/client-mode";	# client mode of xcept for interaction with perl
$timeout     = 30;			# time out for xcept communication
$debug       = 0;			# set true if debug information required
$logfile     = "";			# if set write logging info to this file
$replay      = "";			# if set replay info instead of starting xcept


################################################################################
##  use the following functions as constants for the corresponding
##  BTX symbols.
##
##  &INI
##  &TER

sub INI { "\\*"; }	# BTX-initiator symbol in xcept representation
sub TER { "\\#"; }	# BTX-terminator symbol in xcept representation


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#  if the user has defined a function on_error in main call
#  this function each time an error occures
#
#  error( msg )  -->  -1

sub error {
    if( $errlevel++ > 0 ) {
	die "Double error!! - Exit.\n";
    }
    $error_msg = @_[0];
    $debug && print STDERR "\nPERL: error_msg = $error_msg\n\n";
    if( defined(&main'on_error()) ) {
	&main'on_error($error_msg);
    }
    return -1;
}


################################################################################
##  Initialize package and initiate a connection to the
##  xcept server. Check prompting and version of client mode etc.
##
##  initiate()  -->  0/-1

sub initiate {
    local( $server_exec );
    local( $i );

    if( $replay ne "" ) {
	$debug && print STDERR "PERL: Replaying '$replay'.\n";
	open( XCEPT_S1, "<$replay" ) || &error( "Cannot open '$replay': $!" );
    }
    else {
	if( $client_mode eq "" ) {
	    return &error( "No client script for xcept set" );
	}
	$server_exec = "$prog $prog_opts -s $client_mode";

	$debug && print STDERR "PERL: exec( $server_exec )\n";

	if( !socketpair( XCEPT_S1, XCEPT_S2, &PF_UNIX, &SOCK_STREAM, 0 ) ) {
	    return &error( "Creating a socket pair failed: $!" );
	}

	$SIG{'PIPE'} = 'IGNORE';
	if( ($prog_pid = fork()) < 0 ) {
	    return &error( "Fork failed: $!" );
	}

	if( $prog_pid == 0 ) {			# xcept (server) side
	    fcntl(XCEPT_S2,&F_SETFD,0);		# clear close on exec flag
	    if( !open(STDIN,  "<&XCEPT_S2") ) {
		return &error( "Dupping socket to stdin failed: $!" );
	    }
	    if( !open(STDOUT, ">&XCEPT_S2") ) {
		return &error( "Dupping socket to stdout failed: $!" );
	    }
	    close(XCEPT_S1);
	    close(XCEPT_S2);
	    select(STDOUT); $| = 1;		# make communication unbuffered

	    exec($server_exec);
	    die "Exec failed for: $server_exec\n";
	}

	close(XCEPT_S2);

	if( $logfile && !open( XCEPT_LG, ">$logfile" ) ) {
	    return &error( "Cannot open '$logfile': $!" );
	}
    }
    #
    # set non-blocking reads, since perl reads more than one line
    # on commands like $_=<XCEPT_S1>. Thus you have first to check if
    # there is more data in XCEPT_S1 without blocking if nothing there.
    # if there no more data in XCEPT_S1 (input eq "") than call
    # select.
    #
    fcntl(XCEPT_S1,&F_GETFL,$i);	# non-blocking reads, since perl read more than
    $i |= &FNDELAY;
    fcntl(XCEPT_S1,&F_SETFL,$i);

    select(XCEPT_S1);		# make comm. unbuffered
    $| = 1;
    select(STDOUT);

    if( ($ln = &receive_from_xcept()) eq "" ) {
	return -1;
    }
    if( $ln ne $prompt ) {
	return &error( "Missing initial prompt" );
    }
    return &version();
}


################################################################################
##  shutdown connection to xcept server and terminate the server
##
##  terminate()  --> %

sub terminate {
    &send_to_xcept( "quit" );		# tell xcept to terminate
    close(XCEPT_S1);			# security mode
    $logfile  && close(XCEPT_LG);	# close logfile (if was open)
    $prog_pid && kill( 9, $prog_pid );	# paranoid security mode
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#  Send the specified strings to the xcept server
#
#  send_to_xcept( m1, ... )  -->  0/-1

sub send_to_xcept {
    local( $msg1, $msg2 );

    $msg1 = join( "\n", @_ )."\n";
    $msg2 = join( " ",  @_ )."\n";

    $debug && print STDERR "PERL>>>>XCEPT: $msg2";

    if( $replay ) {
	$msg1 = <XCEPT_S1>;
	if( $msg1 ne $msg2 ) {
	    &error( "Current behavior diverges from replay" );
	}
    }
	
    $logfile && print XCEPT_LG $msg2;
    print XCEPT_S1 $msg1;

    return 0;
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#  Receive exactly one line from the xcept server and
#  return it. Wait only a limited time for input.
#
#  receive_from_xcept()  --> line/""

sub receive_from_xcept {
    local( $nfound, $rin, $rout, $eout, $line );

    $debug && print STDERR "PERL<<<<XCEPT: ";

    while( ($line = <XCEPT_S1>) eq "" ) {	# non-blocking reads
	if( $replay ) {
	    &error( "EOF on replay" );
	    return "";
	}

	$rin  = '';
	vec($rin,fileno(XCEPT_S1),1) = 1;
	$nfound = select( $rout=$rin, undef, $eout=$rin, $timeout );

	if( $nfound < 0 ) {
	    &error( "Select failed: $!" );
	    return "";
	}
	if( $nfound == 0 ) {
	    &error( "Broken communication (TIMEOUT)" );
	    return "";
	}
	if( vec($eout,fileno(XCEPT_S1),1) ) {
	    &error( "Broken communication (Exception)" );
	    return "";
	}
	if( vec($rout,fileno(XCEPT_S1),1) ) {
	    if( ($line = <XCEPT_S1>) eq "" ) {
		&error( "Broken communication (EOF)" );
		return "";
	    }
	    last;
	}
    }
    $debug   && print STDERR  $line;
    $logfile && print XCEPT_LG $line;

    return $line;
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#  Send specified message to xcept server and read in turn from
#  xcept up to the next prompt. All lines with a special prefix
#  are treated as regular reply to the issued command. All other
#  lines are saved and passed to error().
#  n specifies the expected number of regular reply lines.
#  The procedure fails if n doesn't match the replied lines.
#  If n < 0 the check is omitted.
#
#  command( n, m1, [m2, ...] )  --> 0/-1

sub command {
    local( $nlines ) = shift( @_ );
    local( $ln );
    local( $errlines );

    $errlines = "";
    @reply = ();

    if( &send_to_xcept( @_ ) < 0 ) {
	return -1;
    }
    while( ($ln = &receive_from_xcept()) ne "" ) {
	if( $ln eq $prompt ) {
	    last;
	}
	if( $ln =~ s/^$prefix// ){
	    chop $ln;
	    push( @reply, $ln );
	}
	else {
	    $errlines .= "xcept I/O: " . $ln;
	}
    }
    if( $errlines ne "" ) {
	return &error( $errlines );
    }
    if( $nlines >= 0 && $#reply != ($nlines - 1) ) {
	return &error( "Unexpected number of reply lines" );
    }
    return 0;
}

################################################################################
##
##  The rest are procedures relying on command()
##
################################################################################


##  version()  --> 0/-1
##
sub version {
    if( &command( 1, "version" ) < 0 ) {
	return -1;
    }
    if( $reply[0] > $version ) {
	return &error( "Wrong version (found $reply[0], but expecting $version or less)" );
    }
    return 0;
}


##  connect()  -->  0/-1
##
sub connect {
    return &command( -1, "connect" );
}


##  disconnect()  -->  0/-1
##
sub disconnect {
    return &command( 0, "disconnect" );
}


##  status()  -->  "status"/""
##
sub status {
    local( $stat );

    if( &command( 1, "status" ) < 0 ) {
	return "";
    }
    return $reply[0];
}


##  dial( number )  -->  0/-1
##
sub dial {
    return &command( 0, "dial", @_[0] );
}


##  screen()  --> @s[0-23]/()
##
sub screen {
    if( &command( 24, "screen" ) < 0 ) {
	return ();
    }
    return @reply;
}


##  waitdct()  --> 0/-1
##
sub waitdct {
    return &command( 0, "waitdct" );
}


##  diaplay_on( [x-disp-spec] )  --> 0/-1
##
sub display_on {
    if( $_[0] ne "" && &command( 0, "display", $_[0] ) < 0 ) {
	return -1;
    }
    return &command( 0, "onx" );
}


##  display_off()  --> 0/-1
##
sub display_off {
    return &command( 0, "offx" );
}


##  playback( filename )  -->  0/-1
##
sub playback {
    return &command( 0, "playback", $_[0] );
}


##  debug_xcept_client_mode()  -->  0/-1		defaults to on
##  debug_xcept_client_mode( "on"/"off" )  -->  0/-1
##
sub debug_xcept_client_mode {
    return &command( 0, "debug", $#_ >= 0 ? $_[0] : "on" );
}


##  send( s1, [s2, ...] )  -->  0/-1
##
##  send s1,... to btx server as typed to BTX
##
sub send {
    return &command( 0, "send", join("", @_) );
}


##  next()  --> 0/-1
##
##  Send terminator and wait for DCT
##  for user convenience (this sequence is often used)
##
sub next {
    return &command( 0, "send", &TER )
	|| &waitdct();
}


## input( prompt, default )  -->  "input\n"/""
##
## let xcept prompt for input (via popup forms if
## X display is on).
##
sub input {
    local( $iprompt, $idefault ) = @_;

    if( &command( 1, "input", $iprompt, $idefault ) ) {
	return "";
    }
    return $reply[0]."\n";
}


## guest_login()  -->  0/-1
##
## Login via guest id
##
sub guest_login {

    return &connect()		# connect to btx server and wait for btx online
        || &next()		# take preprinted guest id
        || &waitdct();		# wait for intro page 2
}


## user_login( loginID, passwd, userID )  -->  0/-1
##
## Login as user.
## login-id  must be a string of exactly 12 digits
## passwd    is a password string of up to 8 characters
## user-id   must be a string of a most 4 digits
##           (if omitted defaults to "\#" = preprinted userID)
## btxnr     BTX-number (string of up to ?? digits)
##           (if omited defaults to "\#" = preprinted number)
##
sub user_login {

    local( $loginID, $passwd, $userID, $btxnr ) = @_;
    local( $_ );

    # check validy of parameters
    #
    if( $loginID !~ /^[0-9]{12}$/ ) {
	return &error( "Badly formed login ID" );
    }
    if( $passwd !~ /^.{4,8}$/ ) {
	return &error( "Badly formed password" );
    }
    if( $userID ne "" && $userID !~ /^[1-9][0-9]{0,3}$/ ) {
	return &error( "Badly formed user ID" );
    }
    if( $btxnr ne "" && $btxnr !~ /^[0-9]+$/ ) {
	return &error( "Badly formed BTX number" );
    }

    # add #'s if necessary
    #
    if( length($userID) <  4 ) { $userID .= &TER; }
    if( length($passwd) <  8 ) { $passwd .= &TER; }
    if( length($btxnr)  < 12 ) { $btxnr  .= &TER; }

    # now let's try it
    #
    if( &connect()		# connect to btx server and wait for btx online
     || &send( $loginID )	# if ID is wrong BTX disconnects!
     || &waitdct() ) {		# wait for next intro page or error page
	return -1;
    }
    if( &status() !~ /^online/i ) {
	return &error( "Login procedure failed" );
    }
    if( &send( $btxnr )
    ||  &waitdct()
    ||  &send( $userID )
    ||  &waitdct()
    ||  &send( $passwd )
    ||  &waitdct() ) {
	return -1;
    }
    $_ = join("", &screen());
    if( !/Sie benutzten .* zuletzt/ ) {
	return &error( "Login procedure failed" );
    }

    return 0;
}

1;

